mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +347,206 @@ propagateUnmergeVectorization(T &&dimAndLength,
   return result;
 }
 
+// Key data structures for vectorization analysis
+using DimToMergeMap =
+    llvm::DenseMap<uint32_t,
+                   std::tuple<TransformMapAttr, TransformAttr, size_t>>;
+
+using ContiguousMergesMap =
+    llvm::DenseMap<std::pair<TransformMapAttr, TransformAttr>,
+                   SmallVector<SmallVector<uint32_t>>>;
+
+void findCountiguousGroupsUnmerge(const ArrayRef<uint32_t> upperDims,
+                                  const ArrayRef<int64_t> params,
+                                  DimToMergeMap &dimToMerge,
+                                  ContiguousMergesMap &contiguousGroups) {
+  size_t i = 0;
+  while (i < upperDims.size()) {
+    // The i-th dimension (i.e., dimension upperDims[i])
+    // uses params[i]
+
+    auto keyI = dimToMerge[upperDims[i]];
+    auto mergePairI = std::make_pair(std::get<0>(keyI), std::get<1>(keyI));
+    if (!mergePairI.second) {
+      i++;
+      continue;
+    }
+
+    SmallVector<uint32_t> groupCandidate;
+    SmallVector<size_t> dimPosition;
+    for (size_t j = i; j < upperDims.size(); j++) {
+
+      // Unit lengths don't affect the mergeability logic
+      if (params[j] == 1) {
+        continue;
+      }
+
+      auto keyJ = dimToMerge[upperDims[j]];
+      auto mergePairJ = std::make_pair(std::get<0>(keyJ), std::get<1>(keyJ));
+
+      if (!mergePairJ.second) {
+        break;
+      }
+      auto mergePairJParams = mergePairJ.second.getParams();
+      auto mergePairJDims = mergePairJ.second.getLowerDims();
+      size_t posJ = std::get<2>(keyJ);
+
+      // a) Dimensions need to come from the same merge
+      // b) Unmerge parameters need to match merge parameters
+      if (mergePairJ != mergePairI || params[j] != mergePairJParams[posJ]) {
+        break;
+      }
+
+      groupCandidate.push_back(mergePairJDims[posJ]);
+      dimPosition.push_back(posJ);
+    }
+
+    i += std::max(size_t(1), groupCandidate.size());
+
+    // Update the result with the current group for the mergePair key
+    if (groupCandidate.size() > 1 &&
+        std::is_sorted(dimPosition.begin(), dimPosition.end())) {
+      contiguousGroups[mergePairI].push_back(groupCandidate);
+    }
+  }
+}
+
+// This is an analysis pass to group the lower dimensions of a Merge
+// transformation into contiguous groups. E.g., if we have a Merge{8,8,3} [0] ->
+// [0,2,3] and we know that [2,3] are contiguous in the final representation, we
+// can split the merge group in
+// [[0] [2,3]]. This information will be used by the vectorizer. E.g., the
+// vectorizer can fulfill a vectorization by 4, since 8*3=24 is a multiple of 4.
+// In other words, every group of dimensions is treated as a single group
+ContiguousMergesMap findContiguousGroups(ArrayAttr transforms,
+                                         ArrayRef<int64_t> outputShape) {
+  // Transform table
```

Comment:
Comment addition: Will be overwritten after processing each transform_map.

(That way we're all clear that this could've been factored out into a separate function but wasn't)